### PR TITLE
Add kubelet_node_name metrics

### LINKED ISSUE #72715: Have kubelet export metric containing its own name
**What would you like to be added**:

It would be useful if the kubelet exported a metric that contained its own name as a label, for instance:

`kubelet_node_name{job="kubelet", instance="10.135.73.142:10250", name="10.135.73.142"}`.

This name should be consistent with the output of `kubectl get nodes`.

**Why is this needed**:

This would allow us to join the kubelet metrics against metrics exported by kube-state-metrics.

An example of this can be seen in https://github.com/kubernetes-monitoring/kubernetes-mixin/pull/135.  We've arranged in our cluster for the `instance` labels to be set to the node's name, but this is controlled by Prometheus and depends on your scrape & relabelling config.  As such, in general we can't rely on this to be the case - for instance, with the prometheus-operator the `instance` label is a IP:Port combo.

/cc @metalmatze @brancz

### PR DESCRIPTION
**Pull request by danielqsj**
**What type of PR is this?**

/sig instrumentation
/kind feature

**What this PR does / why we need it**:

This would allow us to join the kubelet metrics against metrics exported by kube-state-metrics.

**Which issue(s) this PR fixes**:

Fixes #72715

**Special notes for your reviewer**:

example: 
```
# HELP kubelet_node_name The node's name. The count is always 1.
# TYPE kubelet_node_name gauge
kubelet_node_name{node="nodeName-1"} 1
```

**Does this PR introduce a user-facing change?**:

```release-note
Add kubelet_node_name metrics.
```

### CODE REVIEW

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1303,6 +1303,7 @@ func (kl *Kubelet) initializeModules() error {
 		collectors.NewVolumeStatsCollector(kl),
 		collectors.NewLogMetricsCollector(kl.StatsProvider.ListPodStats),
 	)
+	metrics.SetNodeName(kl.nodeName)
```
Comment (brancz) [inline pkg/kubelet/kubelet.go]: can a node name be changed without a process restart? /cc @derekwaynecarr

**Review by brancz — COMMENTED** [review]
just one comment I'm wondering about, otherwise this would lgtm

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (danielqsj) [reply pkg/kubelet/kubelet.go]: I do not find the place to change node name in code. But not sure.

Comment (derekwaynecarr) [reply pkg/kubelet/kubelet.go]: @brancz no.

Comment (brancz) [reply pkg/kubelet/kubelet.go]: Thanks for the confirmation!

**Comment by tomwilkie** [comment]
Thanks @danielqsj!

**Comment by danielqsj** [comment]
@random-liu could you help review this, thanks

**Comment by danielqsj** [comment]
@thockin @Random-Liu @dchen1107 @vishh @yujuhong  if you have time, can you help review this? Thanks

**Comment by vishh** [comment]
It feels strange to expose node name a constant as a time series metric. Instead why not expose node name as a label for all metrics exposed by the kubelet and join using labels? 
Along the way, kube-state-metrics should export `node-name` in addition to node ip address based on the [published schema](https://github.com/kubernetes/kube-state-metrics/blob/master/Documentation/node-metrics.md)

**Comment by brancz** [comment]
The reasoning as described in https://github.com/kubernetes/kubernetes/issues/72715 is to be able to get the `instance` label of a node merged onto kube-state-metrics metrics.

**Comment by vishh** [comment]
Hmm, isn't adding a `label` different from exposing a dedicated `metric`
which this PR happens to do (or am I misinterpreting the code?

**Comment by brancz** [comment]
This is about allowing consistent querying (more specifically joining) no matter what your Prometheus relabeling config is (this enables alert/dashboard sharing without necessity to configure things, and a single metric is extremely cheap so it's worth it).

**Comment by vishh** [comment]
I'm still lost trying to understand how creating a separate metric is
better for joining compared to ensuring all node metrics expose the node
name (in addition to the dubious ip:host pair that's available today in
some environments).
Can someone post an example query that attempts a join which doesn't work
today?

**Comment by brancz** [comment]
Queries such as what is described in https://github.com/kubernetes-monitoring/kubernetes-mixin/pull/135. Not having this metric, makes Kubernetes dictate how the Prometheus configuration _must_ be, when Kubernetes should try to let users express their opinion in their configuration and not stand in their way.

**Comment by derekwaynecarr** [comment]
missed @vishh comment, holding until he has a chance to ack @brancz response.

**Comment by vishh** [comment]
@brancz let's put aside philosophical stance aside for a bit and I'd appreciate if you can help me understand this problem a bit more in detail!

https://github.com/kubernetes-monitoring/kubernetes-mixin/pull/135#issuecomment-452621738 states that there is a mismatch in labels between two metrics. `kubelet_running_pod_count` is exposed by kubelet & `kube_node_status_capacity` is exposed by kube-state-metrics. It looks like there is a mismatch in labels. 
From the code, kubelet doesn't seem to export any standard labels for most metrics, and kube-state-metrics seems to export node address as a labels which isn't a good for user experience.
I'm asking if we can the do the following, and if it would help address the underlying problem:
1. Have kubelet expose a standard set of labels for all metrics - we can start with `node-name` for start with, but I'd expect other ones to show up like kernel version, OS type/distro & version, k'let version, etc. 
2. Extend kube-state-metrics to expose `node name` in addition to node address as a label.

**Comment by brancz** [comment]
What you're suggesting is almost, the case :slightly_smiling_face: .

1. Adding a label to all metrics goes against best practices, if that's what a user wants, this should be a label on the target (which would enforce a certain Prometheus configuration, which is what we're trying to prevent with this), as all labels of a target are added to the metrics at ingestion time. All we're looking for is 1 metric on the kubelet that can, regardless of Prometheus configuration, reliably be joined on. All of the other labels you're suggesting, can then reliably be joined onto kubelet metrics via the metrics from kube-state-metrics :slightly_smiling_face: .
2. I think there was a misunderstanding here, kube-state-metrics already exposes the `node` label for metrics about node objects, no address.

**Comment by vishh** [comment]
Gotcha. This is essentially an optimization for not having to expose the
same set of labels for every metric. Why is adding a label to all metrics
not considered a "best practice"? Is it coz the prom exposition format is
expensve?
This optimization is opaque, where folks consuming prom metrics and pushing
to other monitoring stacks may not be able to discover this unless it's
well documented & applies across components.

As for `kube-state-metrics` the documentation may need an update then coz
it currently states only node-address is exported -
https://github.com/kubernetes/kube-state-metrics/blob/master/Documentation/node-metrics.md

**Comment by brancz** [comment]
That would be wasted bytes transferred over the wire, when Prometheus can just as well attach the label at ingestion time (if the label in question is static, which is what we're talking about here). Even if pushed, the metric added here is useful for the very same reason mentioned above.

You're right the documentation on kube-state-metrics is incorrect, that should not say node-address, but node-name. I'll fix that.

Let me know if there is anything else I can clarify :slightly_smiling_face: .

**Comment by vishh** [comment]
I'd appreciate if there is some clear documentation on the use of a single
metric to expose metadata relevant to a monitored object/target and some
plans to enforce that across the board to avoid having to know how each k8s
components behaves from a metrics labeling perspective. This can happen
outside of this PR, but having a confirmation on that (and an issue filed)
would be greatly appreciated.

/hold cancel

**Comment by brancz** [comment]
Glad everything got cleared up! This type of metric would be an "info" metric, which is already described in the instrumentation guidelines under normalization :slightly_smiling_face: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/instrumentation.md#normalization

/retest

**Pull request merged**